File: tests/MLPKrigingTest.cpp
#define _USE_MATH_DEFINES  // required for Visual Studio
/**
 * @file MLPKrigingTest.cpp
 * @brief Tests for the MLPKriging class (joint-MLP Deep Kernel Learning).
 */

#include "libKriging/MLPKriging.hpp"

#include <cassert>
#include <cmath>
#include <iostream>

using namespace libKriging;

static double f1d(double x) {
  return 1.0 - 0.5 * (std::sin(12.0 * x) / (1.0 + x) + 2.0 * std::cos(7.0 * x) * std::pow(x, 5) + 0.7);
}

static double f2(double x1, double x2) {
  return std::sin(3.0 * x1) + std::cos(4.0 * x2) + 0.5 * x1 * x2;
}

// ==========================================================================
//  Test 1: 1D + 2D fit/predict/simulate/update
// ==========================================================================
static void test_mlp_kriging_basic() {
  std::cout << "=== Test 1: MLPKriging (Deep Kernel Learning) ===" << std::endl;

  // --- 1a: 1D function ---
  {
    arma::vec X_train = arma::linspace(0.01, 0.99, 10);
    arma::mat X_mat(X_train.n_elem, 1);
    X_mat.col(0) = X_train;
    arma::vec y(X_train.n_elem);
    for (arma::uword i = 0; i < X_train.n_elem; ++i)
      y(i) = f1d(X_train(i));

    MLPKriging model({16, 8}, 2, "selu", "gauss");
    model.fit(y, X_mat, "constant", true, "Adam", "LL", {{"max_iter_adam", "300"}});

    std::cout << model.summary();

    auto [mean_train, stdev_train, _c1b, _md1b, _sd1b] = model.predict(X_mat, true, false);
    double max_err = arma::max(arma::abs(mean_train - y));
    std::cout << "  [1D] Max train interp error:  " << max_err << std::endl;

    arma::vec x_pred = arma::linspace(0.01, 0.99, 50);
    arma::mat xp(50, 1);
    xp.col(0) = x_pred;
    auto [mean, stdev, _c2b, _md2b, _sd2b] = model.predict(xp, true, false);
    arma::vec ytrue(50);
    for (arma::uword i = 0; i < 50; ++i)
      ytrue(i) = f1d(x_pred(i));
    double rmse = std::sqrt(arma::mean(arma::square(mean - ytrue)));
    std::cout << "  [1D] RMSE on dense grid:      " << rmse << std::endl;
  }

  // --- 1b: Branin 2D ---
  {
    auto branin = [](double x1, double x2) -> double {
      double a = 1.0, b = 5.1 / (4.0 * M_PI * M_PI);
      double c = 5.0 / M_PI, r = 6.0, s = 10.0, t = 1.0 / (8.0 * M_PI);
      return a * std::pow(x2 - b * x1 * x1 + c * x1 - r, 2) + s * (1.0 - t) * std::cos(x1) + s;
    };

    arma::uword n = 20;
    arma::arma_rng::set_seed(77);
    arma::mat X = arma::randu<arma::mat>(n, 2);
    arma::vec y(n);
    for (arma::uword i = 0; i < n; ++i)
      y(i) = branin(X(i, 0) * 15.0 - 5.0, X(i, 1) * 15.0);

    MLPKriging model({32, 16}, 3, "selu", "matern5_2");
    model.fit(y, X, "constant", true, "Adam", "LL", {{"max_iter_adam", "300"}});

    std::cout << model.summary();

    arma::mat X_test = arma::randu<arma::mat>(10, 2);
    auto [mean, stdev, _c3b, _md3b, _sd3b] = model.predict(X_test, true, false);
    assert(mean.n_elem == 10);
    assert(stdev.n_elem == 10);
    assert(mean.is_finite());
    assert(stdev.is_finite());

    std::cout << "  [2D] Predicted mean range:  [" << arma::min(mean) << ", " << arma::max(mean) << "]" << std::endl;
    std::cout << "  [2D] Predicted stdev range: [" << arma::min(stdev) << ", " << arma::max(stdev) << "]" << std::endl;

    arma::mat sims = model.simulate(20, 42, X_test);
    assert(sims.n_rows == 10 && sims.n_cols == 20);
    assert(sims.is_finite());

    arma::mat X_new = arma::randu<arma::mat>(3, 2);
    arma::vec y_new(3);
    for (arma::uword i = 0; i < 3; ++i)
      y_new(i) = branin(X_new(i, 0) * 15.0 - 5.0, X_new(i, 1) * 15.0);
    model.update(y_new, X_new);
    std::cout << "  [2D] n after update: " << model.y().n_elem << std::endl;
    assert(model.y().n_elem == n + 3);
  }

  std::cout << "  PASSED\n" << std::endl;
}

// ==========================================================================
//  Test 2: Derivative check vs finite differences
// ==========================================================================
static void check_deriv_vs_fd(const MLPKriging& model,
                              const arma::mat& X_new,
                              const std::vector<arma::uword>& check_dims,
                              const std::string& label,
                              double h = 1e-6,
                              double rel_tol = 0.2,
                              double abs_tol = 0.1) {
  auto [mean, stdev, cov, mean_deriv, stdev_deriv] = model.predict(X_new, true, false, true);

  const arma::uword n_n = X_new.n_rows;
  for (arma::uword i = 0; i < n_n; ++i) {
    for (arma::uword dim : check_dims) {
      arma::mat Xp = X_new;
      arma::mat Xm = X_new;
      Xp(i, dim) += h;
      Xm(i, dim) -= h;
      auto [mp, sp, _cp, _mdp, _sdp] = model.predict(Xp, true, false);
      auto [mm, sm, _cm, _mdm, _sdm] = model.predict(Xm, true, false);

      double dmean_num = (mp(i) - mm(i)) / (2.0 * h);
      double dstdev_num = (sp(i) - sm(i)) / (2.0 * h);
      double dmean_ana = mean_deriv(i, dim);
      double dstdev_ana = stdev_deriv(i, dim);

      double mean_err = std::abs(dmean_num - dmean_ana) / (std::abs(dmean_num) + abs_tol);
      double stdev_err = std::abs(dstdev_num - dstdev_ana) / (std::abs(dstdev_num) + abs_tol);

      if (mean_err > rel_tol) {
        std::cerr << "  [" << label << "] mean deriv mismatch at i=" << i << " dim=" << dim << ": num=" << dmean_num
                  << " ana=" << dmean_ana << std::endl;
        assert(false);
      }
      if (stdev_err > rel_tol) {
        std::cerr << "  [" << label << "] stdev deriv mismatch at i=" << i << " dim=" << dim << ": num=" << dstdev_num
                  << " ana=" << dstdev_ana << std::endl;
        assert(false);
      }
    }
  }
  std::cout << "  [" << label << "] derivative check PASSED" << std::endl;
}

static void test_predict_derivative() {
  std::cout << "=== Test 2: MLPKriging predict derivative vs finite differences ===" << std::endl;

  arma::arma_rng::set_seed(88);
  const arma::uword n = 20;
  arma::mat X = arma::randu<arma::mat>(n, 2);
  arma::vec y(n);
  for (arma::uword i = 0; i < n; ++i)
    y(i) = f2(X(i, 0), X(i, 1));

  MLPKriging model({16, 8}, 2, "selu", "gauss");
  model.fit(y, X, "constant", false, "Adam", "LL", {{"max_iter_adam", "100"}});

  arma::mat X_new = arma::randu<arma::mat>(5, 2);
  check_deriv_vs_fd(model, X_new, {0, 1}, "MLPKriging 2D gauss");

  std::cout << "  PASSED\n" << std::endl;
}

// ==========================================================================
//  main
// ==========================================================================
int main() {
  try {
    test_mlp_kriging_basic();
    test_predict_derivative();

    std::cout << "============================================\n"
              << "  ALL MLPKriging TESTS PASSED\n"
              << "============================================" << std::endl;
  } catch (const std::exception& e) {
    std::cerr << "\n*** TEST FAILED: " << e.what() << " ***" << std::endl;
    return 1;
  }
  return 0;
}

File: bindings/Julia/jlibkriging/tests/mlp_kriging_test.jl
using Test
using jlibkriging

f_test(x) = 1.0 - 0.5 * (sin(12.0 * x) / (1.0 + x) + 2.0 * cos(7.0 * x) * x^5 + 0.7)

function branin(x1, x2)
    a, b, cc = 1.0, 5.1 / (4 * pi^2), 5.0 / pi
    r, s, t = 6.0, 10.0, 1.0 / (8 * pi)
    return a * (x2 - b * x1^2 + cc * x1 - r)^2 + s * (1 - t) * cos(x1) + s
end

@testset "MLPKriging" begin

    @testset "1D gauss" begin
        X = reshape(collect(range(0.01, 0.99; length=10)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = MLPKriging(y, X, [16, 8], 2;
                       activation="selu", kernel="gauss",
                       normalize=true,
                       parameters=Dict("max_iter_adam" => "300"))

        s = jlibkriging.summary(k)
        @test isa(s, String) && length(s) > 0

        result = predict(k, X)
        @test length(result.mean) == 10
        @test all(isfinite.(result.mean))

        x_pred = reshape(collect(range(0.01, 0.99; length=50)), :, 1)
        result2 = predict(k, x_pred)
        y_true = [f_test(x) for x in x_pred[:, 1]]
        rmse = sqrt(sum((result2.mean .- y_true).^2) / length(y_true))
        @test isfinite(rmse)

        @test kernel(k) == "gauss"
        @test activation(k) == "selu"
        @test get_hidden_dims(k) == [16, 8]
        @test feature_dim(k) == 2
        @test get_sigma2(k) > 0
        @test isfinite(log_likelihood(k))
        @test is_fitted(k)
    end

    @testset "Branin 2D matern" begin
        n = 20
        rng_vals = [0.37454, 0.95071, 0.73199, 0.59866, 0.15601,
                    0.15599, 0.05808, 0.86617, 0.70807, 0.02058,
                    0.96991, 0.83244, 0.21233, 0.18183, 0.18340,
                    0.30424, 0.52475, 0.43194, 0.29122, 0.61185,
                    0.13949, 0.29214, 0.36636, 0.45607, 0.78518,
                    0.19968, 0.51423, 0.59242, 0.04646, 0.60754,
                    0.17052, 0.06505, 0.94889, 0.96563, 0.80840,
                    0.30462, 0.09767, 0.68423, 0.44015, 0.12203]
        X = reshape(rng_vals, n, 2)
        y = [branin(X[i, 1] * 15 - 5, X[i, 2] * 15) for i in 1:n]

        k = MLPKriging(y, X, [32, 16], 3;
                       activation="selu", kernel="matern5_2",
                       normalize=true,
                       parameters=Dict("max_iter_adam" => "300"))

        X_test = reshape([0.5, 0.3, 0.8, 0.1, 0.6,
                          0.4, 0.7, 0.2, 0.9, 0.5,
                          0.2, 0.8, 0.4, 0.9, 0.1,
                          0.6, 0.3, 0.7, 0.5, 0.8], 10, 2)

        result = predict(k, X_test)
        @test length(result.mean) == 10
        @test all(isfinite.(result.mean))
        @test all(isfinite.(result.stdev))

        sims = simulate(k, 20, 42, X_test)
        @test size(sims) == (10, 20)
        @test all(isfinite.(sims))
    end

    @testset "Update" begin
        X = reshape(collect(range(0.01, 0.99; length=10)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = MLPKriging(y, X, [16, 8], 2;
                       activation="selu", kernel="gauss",
                       parameters=Dict("max_iter_adam" => "100"))

        X_test = reshape(collect(range(0.05, 0.95; length=5)), :, 1)
        p_before = predict(k, X_test)

        X_new = reshape([0.15, 0.35, 0.55], :, 1)
        y_new = [f_test(x) for x in X_new[:, 1]]
        update!(k, y_new, X_new)

        p_after = predict(k, X_test)
        @test p_before.mean != p_after.mean
    end

    @testset "Predict with derivatives" begin
        n = 20
        X = reshape(collect(range(0.01, 0.99; length=n)), :, 1)
        y = [sin(3.0 * x) + 0.5 * x for x in X[:, 1]]

        k = MLPKriging(y, X, [16, 8], 2;
                       activation="selu", kernel="gauss",
                       parameters=Dict("max_iter_adam" => "100"))

        X_new = reshape([0.1, 0.3, 0.5, 0.7, 0.9], :, 1)
        result = predict(k, X_new; return_deriv=true)
        @test size(result.mean_deriv) == (5, 1)
        @test size(result.stdev_deriv) == (5, 1)
        @test all(isfinite.(result.mean_deriv))
        @test all(isfinite.(result.stdev_deriv))
    end

    @testset "Predict with covariance" begin
        n = 20
        X = reshape(collect(range(0.01, 0.99; length=n)), :, 1)
        y = [sin(3.0 * x) + 0.5 * x for x in X[:, 1]]

        k = MLPKriging(y, X, [16, 8], 2;
                       activation="selu", kernel="gauss",
                       parameters=Dict("max_iter_adam" => "100"))

        X_new = reshape([0.1, 0.3, 0.5, 0.7, 0.9], :, 1)
        result = predict(k, X_new; return_stdev=true, return_cov=true)
        @test result.cov !== nothing
        @test size(result.cov) == (5, 5)
        @test all(isfinite.(result.cov))
        @test maximum(abs.(result.cov - result.cov')) < 1e-10
    end

    @testset "Getters" begin
        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = MLPKriging(y, X, [16, 8], 2;
                       activation="selu", kernel="gauss",
                       parameters=Dict("max_iter_adam" => "100"))

        @test size(get_X(k)) == (8, 1)
        @test length(get_y(k)) == 8
        @test length(get_theta(k)) > 0
        @test get_sigma2(k) > 0.0
        @test kernel(k) == "gauss"
        @test activation(k) == "selu"
        @test get_hidden_dims(k) == [16, 8]
        @test feature_dim(k) == 2
        @test is_fitted(k)
    end

    @testset "Log-likelihood" begin
        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = MLPKriging(y, X, [16, 8], 2;
                       activation="selu", kernel="gauss",
                       parameters=Dict("max_iter_adam" => "100"))

        ll = log_likelihood(k)
        @test isfinite(ll)

        theta = get_theta(k)
        ll_res = log_likelihood_fun(k, theta)
        @test isfinite(ll_res.ll)
        @test ll_res.grad === nothing

        ll_res_grad = log_likelihood_fun(k, theta; return_grad=true)
        @test isfinite(ll_res_grad.ll)
        @test length(ll_res_grad.grad) == length(theta)
    end

    @testset "Empty constructor then fit" begin
        k = MLPKriging([16, 8], 2; activation="selu", kernel="gauss")
        @test !is_fitted(k)

        X = reshape(collect(range(0.01, 0.99; length=10)), :, 1)
        y = [f_test(x) for x in X[:, 1]]
        fit!(k, y, X; parameters=Dict("max_iter_adam" => "100"))
        @test is_fitted(k)

        result = predict(k, X)
        @test length(result.mean) == 10
        @test all(isfinite.(result.mean))
    end
end
